src/mpid/ch4/netmod/ofi/ofi_finalize.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "ofi_impl.h"
+#include "ofi_noinline.h"
+
+#define MPIDI_OFI_FLUSH_CONTEXT_ID 0xF000
+#define MPIDI_OFI_FLUSH_TAG        1
+
+/* send a dummy message to flush the send queue */
+static int flush_send(int dst)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_Comm *comm = MPIR_Process.comm_world;
+    int data = 0;
+    uint64_t match_bits = MPIDI_OFI_init_sendtag(MPIDI_OFI_FLUSH_CONTEXT_ID,
+                                                 MPIDI_OFI_FLUSH_TAG, MPIDI_OFI_DYNPROC_SEND);
+
+    /* Use the same direct send method as used in establishing dynamic processes */
+    MPIDI_OFI_dynamic_process_request_t req;
+    req.done = 0;
+    req.event_id = MPIDI_OFI_EVENT_DYNPROC_DONE;
+
+    MPIDI_OFI_CALL_RETRY(fi_tsenddata(MPIDI_OFI_global.ctx[0].tx, &data, 4, NULL, 0,
+                                      MPIDI_OFI_av_to_phys(MPIDIU_comm_rank_to_av(comm, dst)),
+                                      match_bits, &req.context), tsenddata, FALSE);
+    MPIDI_OFI_PROGRESS_WHILE(!req.done);
```

Comment:
We have encountered a lot of issues in MPI_Allreduce (used as a barrier) in MPI_Finalize. I see this is an effort to fix it, however, a better fix is to use FI_DELIVERY_COMPLETE (together with FI_SELECTIVE_COMPLETE setup) to make sure a message is delivered in the MPI_Finalize phase. Progress call like this often just wait for send complete, which is not enough to guarantee a clean state before fabrics/endpoints are closed.

Reply:
> a better fix is to use FI_DELIVERY_COMPLETE (together with FI_SELECTIVE_COMPLETE setup) to make sure a message is delivered in the MPI_Finalize phase.)

I guess the idea is -- `FI_DELIVERY_COMPLETE` generates delivery completion, so we could simply wait for sends completed in the finalize? 

That should work if it works. Two questions: first, does it work? From my cursory read, seems `FI_DELIVERY_COMPLETE`'s support is not quite complete, in fact, most providers doesn't support it?  Second question: when it works, what is its impact on performance?

Reply:
> We have encountered a lot of issues in MPI_Allreduce (used as a barrier) in MPI_Finalize.

Oh, I also need point out, that I finally understand that it is not "used as a barrier". The barrier has nothing to do with the purpose. For example, replacing it with `PMI_barrier`, the issue remains.

It is just using the side effects of `MPI_Allreduce` that it places some outgoing send messages for each processes and those messages gets waited on -- exactly what this PR does directly.

Me too get misled by their bad naming choice for so long (maybe it is modified by someone who also lacks clue).